ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudBlobCryptoService.java

```diff
@@ -0,0 +1,158 @@
+   *                    supports int sized sizes
+   *
+   *  encrypted key   - The encrypted key
+   *
+   *  encrypted data  - The encrypted data
+   *
+   *  crc             - The crc of the message
+   *
+   */
+  static class EncryptedKeyAndEncryptedData {
+
+    private static short CURRENT_VERSION = 1;
+    private static int VERSION_FIELD_SIZE = 2;
+    private static int ENCRYPTED_KEY_SIZE_FIELD_SIZE = 4;
+    private static int ENCRYPTED_DATA_SIZE_FIELD_SIZE = 8;
+    static int INITIAL_MESSAGE_LENGTH =
+        VERSION_FIELD_SIZE + ENCRYPTED_KEY_SIZE_FIELD_SIZE + ENCRYPTED_DATA_SIZE_FIELD_SIZE;
+    private static int CRC_FIELD_LENGTH = 8;
+    static int OVERHEAD_LENGTH = INITIAL_MESSAGE_LENGTH + CRC_FIELD_LENGTH;
+
+    private ByteBuffer encryptedKey;
+    private ByteBuffer encryptedData;
+
+    EncryptedKeyAndEncryptedData(ByteBuffer encryptedKey, ByteBuffer encryptedData) {
+      this.encryptedKey = encryptedKey;
+      this.encryptedData = encryptedData;
+    }
+
+    static void serialize(ByteBuffer outputBuffer, EncryptedKeyAndEncryptedData encryptedKeyAndEncryptedData) {
+      int startOffset = outputBuffer.position();
+      outputBuffer.putShort(CURRENT_VERSION);
+      int encryptedKeySize = encryptedKeyAndEncryptedData.encryptedKey.array().length;
+      int encryptedDataSize = encryptedKeyAndEncryptedData.encryptedData.array().length;
+      outputBuffer.putInt(encryptedKeySize);
+      outputBuffer.putLong(encryptedDataSize);
+      outputBuffer.put(encryptedKeyAndEncryptedData.encryptedKey);
+      outputBuffer.put(encryptedKeyAndEncryptedData.encryptedData);
+      Crc32 crc = new Crc32();
+      crc.update(outputBuffer.array(), startOffset, INITIAL_MESSAGE_LENGTH + encryptedKeySize + encryptedDataSize);
+      outputBuffer.putLong(crc.getValue());
+    }
+
+    static EncryptedKeyAndEncryptedData deserialize(ByteBuffer buffer) throws GeneralSecurityException {
+      int startOffset = buffer.position();
+      Crc32 crc = new Crc32();
+      short version = buffer.getShort();
+      if (version != 1) {
+        throw new GeneralSecurityException("Unrecognized version");
+      }
+      int encryptedKeySize = buffer.getInt();
+      int encryptedDataSize = (int) buffer.getLong();
+      if (encryptedKeySize < 0) {
+        throw new GeneralSecurityException("Encrypted key size is less than 0");
+      }
+      if (encryptedDataSize < 0) {
+        throw new GeneralSecurityException("Encrypted data size is less than 0");
+      }
+      byte[] encryptedKey = new byte[encryptedKeySize];
+      byte[] encryptedData = new byte[encryptedDataSize];
+      buffer.get(encryptedKey);
+      buffer.get(encryptedData);
+      crc.update(buffer.array(), startOffset, INITIAL_MESSAGE_LENGTH + encryptedKeySize + encryptedDataSize);
+      long expectedCrc = buffer.getLong();
+      long actualCrc = crc.getValue();
+      if (actualCrc != expectedCrc) {
+        throw new GeneralSecurityException("Encrypted blob is corrupt");
```

Comment:
Include expected and actual crc in message?

Reply:
sure, added